langchain/src/callbacks/promises.ts

```diff
@@ -0,0 +1,34 @@
+import PQueueMod from "p-queue";
+
+let queue: typeof import("p-queue")["default"]["prototype"];
+
+function createQueue() {
+  const PQueue = "default" in PQueueMod ? PQueueMod.default : PQueueMod;
+  return new PQueue({
+    autoStart: true,
+    concurrency: 1,
+  });
+}
+
+/**
+ * Consume a promise, either adding it to the queue or waiting for it to resolve
+ * @param promise Promise to consume
+ * @param wait Whether to wait for the promise to resolve or resolve immediately
+ */
+export async function consumeCallback<T>(
+  promiseFn: () => Promise<T> | T | void,
+  wait: boolean
+): Promise<void> {
+  if (wait === true) {
+    await promiseFn();
+  } else {
+    if (typeof queue === "undefined") {
+      queue = createQueue();
```

Comment:
No strong feelings but we could just make this a global singleton?

Reply:
It's to avoid the side effect the bundlers would complain about in that case